system.clients: add `client_options` map column

vladzcloudius opened the pull request:
This pull request introduces a new caching mechanism for client options in the Alternator and transport layers, refactors how client metadata is stored and accessed, and extends the `system.clients` virtual table to surface richer client information. The changes improve efficiency by deduplicating commonly used strings (like driver names/versions and client options), and ensure that client data is handled in a way that's safe for cross-shard access. Additionally, the test suite and virtual table schema are updated to reflect the new client options data.

**Caching and client metadata refactoring:**

###### The motivation
* The largest and most repeatable items in the connection state before this PR were a `driver_name` and a `driver_version` which were stored as an `sstring` object which means that the corresponding memory consumption was 16 bytes per each such value at least (the smallest size of the `seastar`'s `sstring` object) **per-connection**. In reality the driver name is usually longer than 15 characters, e.g. "ScyllaDB Python Driver" is 23 characters and this is not the longest driver name there is. In such cases the actual memory usage of a corresponding `sstring` object jumps to 8 + 4 + 1 + (string length, 23 in our example) + 1.
So, for "ScyllaDB Python Driver" it would be 37 bytes (in reality it would be a bit more due to natural alignment of other allocations since the `contents` size is not well aligned (13 bytes), but let's ignore this for now).
* These bytes add up quickly as there are more connections and, sometimes we are talking about millions of connections per-shard.
* Using a smart pointer (`lw_shared_ptr`) referencing a corresponding cached value will effectively reduce the per-connection memory usage to be 8 bytes (a size of a pointer on 64-bit CPU platform) for each such value. While storing a corresponding `sstring` value only once.
* This will would reduce the "variable" (per-connection) memory usage by **at least 50%**. And in case of "ScyllaDB Python Driver" driver version - by 78%!
* And all this for a price of a single `loading_shared_values` object **per-shard** (implements a hash table) and a minor overhead for each value **stored** in it.

###### The implementation
* Introduced a new cache type (`client_options_cache_type`) for deduplicating and sharing client option strings, and refactored `client_data`, `client_state`, and related classes to use `foreign_ptr<std::unique_ptr<client_data>>` and cached entry types for fields like driver name, driver version, and client options. (`client_data.hh`, `service/client_state.hh`, `alternator/server.hh`, `alternator/controller.hh`, `transport/controller.hh`, `transport/protocol_server.hh`) [[1]](diffhunk://#diff-664a3b19e905481bdf8eb3843fc4d34691067bb97ab11cfd6e652e74aac51d9fR33-R36) [[2]](diffhunk://#diff-664a3b19e905481bdf8eb3843fc4d34691067bb97ab11cfd6e652e74aac51d9fL40-R56) [[3]](diffhunk://#diff-daadce1a2de3667511e59558f3a8f077b5ee30a14bcc6a99d588db90d0fcd2bdL105-R107) [[4]](diffhunk://#diff-daadce1a2de3667511e59558f3a8f077b5ee30a14bcc6a99d588db90d0fcd2bdL154-R182) [[5]](diffhunk://#diff-5fce246edf5abffb2351bd02e2eb1e9850880f7a00607ccaa90c3eee7ef57c6bL91-R92) [[6]](diffhunk://#diff-5fce246edf5abffb2351bd02e2eb1e9850880f7a00607ccaa90c3eee7ef57c6bL110-R111) [[7]](diffhunk://#diff-31730ba8e7374f784a88dc27c1512291cf73b7f24e08768f7466a3c8cfcc7a1aL96-R96) [[8]](diffhunk://#diff-19a97c0247cc08155ee49b277e43859ca32d6ef8cbff0ed7368ec5fa19e0a11eL172-R172) [[9]](diffhunk://#diff-eea7e2db5d799a25e717a72ac8ce5842bd4adb72b694d38d8f47166d9cd926faL356-R356) [[10]](diffhunk://#diff-d0b4ec3a144bbc5dc993866cf0b940850a457ff6156064f7e2b4b10ad0a95fefL80-R80) [[11]](diffhunk://#diff-4293b94c444d9bd5ecd17ce7eda8c00685d35ecf6e07f844efc91a91bbe85be1L46-R48)

* Updated the methods for setting and getting driver name, driver version, and client options in `client_state` to be asynchronous and use the new cache. (`service/client_state.hh`, `service/client_state.cc`) [[1]](diffhunk://#diff-daadce1a2de3667511e59558f3a8f077b5ee30a14bcc6a99d588db90d0fcd2bdL154-R182) [[2]](diffhunk://#diff-99634aae22e2573f38b4e2f050ed2ac4f8173ff27f0ae8b3609d1f0cc1aeb775R347-R362)

**Virtual table and API enhancements:**

* Extended the `system.clients` virtual table schema and implementation to include a new `client_options` column (a map of option key/value pairs), and updated the table population logic to use the new cached types and foreign pointers. (`db/virtual_tables.cc`) [[1]](diffhunk://#diff-05f7bff3edb39fb8759c90b445e860189f2f30e04717ed58bae42716082af3d1R752) [[2]](diffhunk://#diff-05f7bff3edb39fb8759c90b445e860189f2f30e04717ed58bae42716082af3d1L769-R770) [[3]](diffhunk://#diff-05f7bff3edb39fb8759c90b445e860189f2f30e04717ed58bae42716082af3d1L809-R816) [[4]](diffhunk://#diff-05f7bff3edb39fb8759c90b445e860189f2f30e04717ed58bae42716082af3d1L828-R879)

**API and interface changes:**

* Changed the signatures of `get_client_data` methods throughout the codebase to return vectors of `foreign_ptr<std::unique_ptr<client_data>>` instead of plain `client_data` objects, to ensure safe cross-shard access. (`alternator/controller.hh`, `alternator/controller.cc`, `alternator/server.hh`, `alternator/server.cc`, `transport/controller.hh`, `transport/protocol_server.hh`) [[1]](diffhunk://#diff-31730ba8e7374f784a88dc27c1512291cf73b7f24e08768f7466a3c8cfcc7a1aL96-R96) [[2]](diffhunk://#diff-19a97c0247cc08155ee49b277e43859ca32d6ef8cbff0ed7368ec5fa19e0a11eL172-R172) [[3]](diffhunk://#diff-5fce246edf5abffb2351bd02e2eb1e9850880f7a00607ccaa90c3eee7ef57c6bL110-R111) [[4]](diffhunk://#diff-a7e2cda866c03a75afcf3b087de1c1dcd2e7aa996214db67f9a11ed6451e596dL988-R995) [[5]](diffhunk://#diff-eea7e2db5d799a25e717a72ac8ce5842bd4adb72b694d38d8f47166d9cd926faL356-R356) [[6]](diffhunk://#diff-d0b4ec3a144bbc5dc993866cf0b940850a457ff6156064f7e2b4b10ad0a95fefL80-R80) [[7]](diffhunk://#diff-4293b94c444d9bd5ecd17ce7eda8c00685d35ecf6e07f844efc91a91bbe85be1L46-R48)

**Testing and validation:**

* Updated the Python test for the `system.clients` table to verify the new `client_options` column and its contents, ensuring that driver name and version are present in the options map. (`test/cqlpy/test_virtual_tables.py`) [[1]](diffhunk://#diff-6dd8bd4a6a82cd642252a29dc70726f89a46ceefb991c3e63fc67e283f323f03R79) [[2]](diffhunk://#diff-6dd8bd4a6a82cd642252a29dc70726f89a46ceefb991c3e63fc67e283f323f03R88-R90)

avikivity commented:
Missing documention in cql-extensions.md.

```
    0x04: Custom payload flag. For a request or response frame, this indicates
          that a generic key-value custom payload for a custom QueryHandler
          implementation is present in the frame. Such a custom payload is simply
          ignored by the default QueryHandler implementation.
          Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
          payload.
```

We might ignore that.

mykaul commented:
I don't know so asking - does it work transparently without a feature flag, across two different Scylla versions (while upgrade is taking place, for example). Since it's a virtual table, perhaps it's OK?

avikivity commented:
> I don't know so asking - does it work transparently without a feature flag, across two different Scylla versions (while upgrade is taking place, for example). Since it's a virtual table, perhaps it's OK?

The custom payload is ignored when not recognized.

tzach commented:
@vladzcloudius can a driver change config in mid-session? if yes, would it resend the DRIVER_CONFIG?
(clearly its a driver, not server, requirement; question is, can it be done on the STARTUP event)

vladzcloudius commented:
> @vladzcloudius can a driver change config in mid-session? if yes, would it resend the DRIVER_CONFIG? (clearly its a driver, not server, requirement; question is, can it be done on the STARTUP event)

`STARTUP` opcode can only be sent once during a specific session according to the CQL protocol spec:

```
4.1.1 STARTUP

  Initialize the connection. The server will respond by either a READY message
  (in which case the connection is ready for queries) or an AUTHENTICATE message
  (in which case credentials will need to be provided using AUTH_RESPONSE).

  This must be the first message of the connection, except for OPTIONS that can
  be sent before to find out the options supported by the server. Once the
  connection has been initialized, a client should not send any more STARTUP
  messages.
```

Do drivers even allow mid-session changes of their configuration? 
If they don't IMO we should not implement it. The value of this is questionable.
And from the perspective of a new featured introduced by this PR it would require us parsing custom payload of every "fast path" opcodes like `EXECUTE` and `QUERY`. 

So, instead we'd rather require re-establishing new sessions for such changes.
The same applies to `COMPRESSION` and/or all `SCYLLA_XXX` options we support.

> Missing documention in cql-extensions.md.
> 
> ```
>     0x04: Custom payload flag. For a request or response frame, this indicates
>           that a generic key-value custom payload for a custom QueryHandler
>           implementation is present in the frame. Such a custom payload is simply
>           ignored by the default QueryHandler implementation.
>           Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
>           payload.
> ```
> 
> We might ignore that.

Actually, that's a good catch.
So, my PR/patch descriptions were misleading. I was not using a custom payload header but rather via an `OPTIONS` section.

And according to a binary protocol spec we should declare each option the server supports in the `SUPPORTED` response: done in `cql_server::connection::make_supported(...)`.

@dkropachev FYI ^

vladzcloudius commented:
> > @vladzcloudius can a driver change config in mid-session? if yes, would it resend the DRIVER_CONFIG? (clearly its a driver, not server, requirement; question is, can it be done on the STARTUP event)
> 
> `STARTUP` opcode can only be sent once during a specific session according to the CQL protocol spec:
> 
> ```
> 4.1.1 STARTUP
> 
>   Initialize the connection. The server will respond by either a READY message
>   (in which case the connection is ready for queries) or an AUTHENTICATE message
>   (in which case credentials will need to be provided using AUTH_RESPONSE).
> 
>   This must be the first message of the connection, except for OPTIONS that can
>   be sent before to find out the options supported by the server. Once the
>   connection has been initialized, a client should not send any more STARTUP
>   messages.
> ```
> 
> Do drivers even allow mid-session changes of their configuration? If they don't IMO we should not implement it. The value of this is questionable. And from the perspective of a new featured introduced by this PR it would require us parsing custom payload of every "fast path" opcodes like `EXECUTE` and `QUERY`.
> 
> So, instead we'd rather require re-establishing new sessions for such changes. The same applies to `COMPRESSION` and/or all `SCYLLA_XXX` options we support.
> 
> > Missing documention in cql-extensions.md.
> > ```
> >     0x04: Custom payload flag. For a request or response frame, this indicates
> >           that a generic key-value custom payload for a custom QueryHandler
> >           implementation is present in the frame. Such a custom payload is simply
> >           ignored by the default QueryHandler implementation.
> >           Currently, only QUERY, PREPARE, EXECUTE and BATCH requests support
> >           payload.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We might ignore that.
> 
> Actually, that's a good catch. So, my PR/patch descriptions were misleading. I was not using a custom payload header but was rather getting a new value via an `OPTIONS` section, similarly to `COMPRESSION` option.
> 
> And according to a binary protocol spec we should declare each option the server supports in the `SUPPORTED` response before expecting it in the `STARTUP`: done in `cql_server::connection::make_supported(...)`.
> <img alt="image" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/6103981/484740647-a7983e4e-48f5-4559-8d1f-c29895c20051.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4NTk3MjUsIm5iZiI6MTc1Njg1OTQyNSwicGF0aCI6Ii82MTAzOTgxLzQ4NDc0MDY0Ny1hNzk4M2U0ZS00OGY1LTQ1NTktOGQxZi1jMjk4OTVjMjAwNTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDNUMDAzMDI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjhlZTMyMGY4OWVhOWI3ZmY1ZWU5NDAxM2YwMzZlMDQzNjdjNzQxOWM0ZTE4YzhiMDVhNGU2ZjFmMmNhZjUxYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gftzIF4IEyRa2FLy7ovijO9UC6k-8fSs2Q-YMfGL6NA">
> 
> I'll add and push. Thanks, @avikivity.
> 
> @dkropachev FYI ^

Done.

vladzcloudius commented on transport/server.cc:
```diff
@@ -943,6 +944,11 @@ future<std::unique_ptr<cql_server::response>> cql_server::connection::process_st
         _client_state.set_driver_name(driver_name_opt->second);
     }
 
+    if (auto driver_config = options.find("DRIVER_CONFIG"); driver_config != options.end()) {
+        clogger.debug("Received DRIVER_CONFIG option: {}", driver_config->second);
+        _client_state.set_driver_config(driver_config->second);
+    }
```
> I would prefer way cassandra implemented it:
> 
> ```
> VIRTUAL TABLE system_views.clients (
>     address inet,
>     port int,
>     client_options frozen<map<text, text>>,
> ```
> 
> Whatever goes into startup options endup in `client_options`. It makes it extendable, if we would want to add more information we would not need to change anything on server side. And we have already implemented feature on dirvers that can benefit from that - [scylladb/scylla-drivers#78](https://github.com/scylladb/scylla-drivers/issues/78)
> 
> In future we can report environment, say JDK version and configuration, os, cpu etc...

Okay, let's parse that JSON. ;)

dkropachev replied:
I’m suggesting it would be better to have a dedicated column that stores all the custom metadata from the STARTUP frame, not only `DRIVER_CONFIG` key.
If not, then we will have to have a separate column for every key, since we have have application information reported (same way it is implemented for cassandra) that does not endup anywhere, we will have to add columns for following STARTUP metadata keys:
APPLICATION_NAME
APPLICATION_VERSION
CLIENT_ID

vladzcloudius commented:
> > @roydahan these look like CI/CD bugs. Are you, guys, looking at it (I saw you had assigned @dkropachev)?
> 
> You probably meant to ping @scylladb/releng
> 
> This specific issue waa fixed. Rebase and rerun.

Thanks. Done.

ptrsmrn commented on client_data.hh:
```diff
@@ -46,6 +46,7 @@ struct client_data {
     std::optional<sstring> ssl_protocol;
```
The commit's msg reads:
> Add a new column to a system.clients table that would contain a new SATARTUP opcode

Did you mean "startup"?

ptrsmrn commented:
Do drivers support this new opcode already? If yes, then please provide tests.

vladzcloudius commented on transport/server.cc:
```diff
@@ -1540,6 +1546,9 @@ std::unique_ptr<cql_server::response> cql_server::connection::make_supported(int
     opts.insert({"CQL_VERSION", cql3::query_processor::CQL_VERSION});
     opts.insert({"COMPRESSION", "lz4"});
     opts.insert({"COMPRESSION", "snappy"});
+    // CLIENT_OPTIONS value is a JSON string that can be used to pass clinet-specific configuration,
+    // e.g. CQL driver configuration.
+    opts.insert({"CLIENT_OPTIONS", ""});
```
> This is not needed, driver will send it anyways.

This is required according to the CQL Binary protocol spec:  see "4.2.4 SUPPORTED" in https://cassandra.apache.org/doc/latest/cassandra/_attachments/native_protocol_v5.html#s4.2.4

dkropachev replied:
Yes, but take a look, they don't follow it:
https://github.com/apache/cassandra-java-driver/blob/1e0eae92efb89df94d6f10a52e8d50864e8ed009/core/src/main/java/com/datastax/oss/driver/internal/core/channel/ProtocolInitHandler.java#L193-L201

Again, I don't mind having it.

dkropachev commented:
> How big this new filed is supposed to be in practice? Above certain threshold it would be wasteful to store 10k times the same string.

We will report it only for control connection.

avikivity commented:
There could be 10k clients connected, easily.

Perhaps deduplicate it via some hash table + weak pointers.

dkropachev commented:
> There could be 10k clients connected, easily.
> 
> Perhaps deduplicate it via some hash table + weak pointers.

Yeah, in some cases we have 1m clients.

mykaul commented:
Please resolve conflicts and let's see how we address the issue of many clients and their configs - I'd like to see this in 2025.4 and that's not too far away.

vladzcloudius commented:
> Please resolve conflicts and let's see how we address the issue of many clients and their configs - I'd like to see this in 2025.4 and that's not too far away.

@avikivity, @mykaul  implementing the whole caching thingy is going to take some time.
How about I use the current naive approach and only address @dkropachev comment but will parametrize the population of the new column with a LiveUpdate parameter which will disable the feature by default?

This way we should be able to get it into 2025.4 and already be able to use it in the field by enabling it on demand.
And I'll send a full version as a followup.

@avikivity will it be acceptable from your perspective?

ruthea commented:
> > > Please resolve conflicts and let's see how we address the issue of many clients and their configs - I'd like to see this in 2025.4 and that's not too far away.
> > 
> > 
> > @avikivity, @mykaul implementing the whole caching thingy is going to take some time. How about I use the current naive approach and only address @dkropachev comment but will parametrize the population of the new column with a LiveUpdate parameter which will disable the feature by default?
> 
> That will make it useless. Even if you enable it later, you have to wait until clients reconnect to actually see anything. And before you do that, you have to carefully check if the number of clients isn't such that it will consume too much memory.
> 
> > This way we should be able to get it into 2025.4 and already be able to use it in the field by enabling it on demand. And I'll send a full implementation that addresses Avi's comment as a followup.
> > @avikivity will it be acceptable from your perspective?

It would be a good first start. Agree that it wouldn't be useful when live updated without a client reconnection, but if I understand the option correctly, all we really need to do is restart one node or perform during a Upgrade/RR (ensuring there aren't too many clients - need to establish some ground rules for that) and we can gain lots of information when trying to troubleshoot issues.

vladzcloudius commented:
> Here's a simple way to memoize the values without weak_ptr, which isn't easy to use:
> 
>     1. Add an unordered_map<sstring, size_t>
> 
>     2. When storing a value, use try_emplace(std::move(value), 0) to insert it into the map if it's not there.
> 
>     3. ++i->second to bump its reference count
> 
>     4. store i into the connection instead of value.
> 
>     5. when destroying the connection, --i->second. If it ends up being 0, erase i.

@avikivity, you've just described the internal implementation of a `shared_ptr`. ;)

avikivity commented:
I'm sure that's not how shared_ptr is implemented.

vladzcloudius commented:
@avikivity @mykaul @dkropachev Guys, I'm in the process of addressing all the comments above.
I have a question: the current schema if `system.clients` is as follows:

```
CREATE TABLE system.clients (
    address inet,
    port int,
    client_type text,
    connection_stage text,
    driver_name text,
    driver_version text,
    hostname text,
    protocol_version int,
    scheduling_group text,
    shard_id int,
    ssl_cipher_suite text,
    ssl_enabled boolean,
    ssl_protocol text,
    username text,
    PRIMARY KEY (address, port, client_type)
) WITH CLUSTERING ORDER BY (port ASC, client_type ASC)
 
```

And columns `driver_name` and `driver_version` are derived from the `options` sent by the driver.
Since I'm going to add the whole map as separate column I thought it makes sense to drop these two columns.
As a result the schema will look like follows:

```
CREATE TABLE system.clients (
    address inet,
    port int,
    client_type text,
    client_options frozen<map<text, text>>,
    connection_stage text,
    hostname text,
    protocol_version int,
    scheduling_group text,
    shard_id int,
    ssl_cipher_suite text,
    ssl_enabled boolean,
    ssl_protocol text,
    username text,
    PRIMARY KEY (address, port, client_type)
```

Any objections?

vladzcloudius commented:
It looks like the Alternator's uses the `driver_name` column to pass info. So, let's keep this one.

avikivity commented:
Don't drop columns, we need to be serious about backwards compatibility.

About the column naming, we need to consider that we'll add configuration sent from the database to the driver. On second thought, that will be in the system.service_levels table, so the names won't clash.

vladzcloudius commented:
I want to optimize the current code that stores many options values for each TCP connection, e.g. a driver name and version. And these values are highly repeatable. As a result Scylla burns a huge amount of memory storing corresponding `sstring`s. 

@avikivity, I'm going to add a corresponding patches that would optimize this for the old code  (CQL and Alternator) to use the same caching technique I've implemented here for `client_options`.

vladzcloudius commented:
> I want to optimize the current code that stores many options values for each TCP connection, e.g. a driver name and version. And these values are highly repeatable. As a result Scylla burns a huge amount of memory storing corresponding `sstring`s.
> 
> @avikivity, I'm going to add corresponding patches that would optimize this for the old code (CQL and Alternator) to use the same caching technique I've implemented here for `client_options`.

Done

vladzcloudius commented:
@scylladbbot trigger-ci

vladzcloudius commented:
```16:31:09  clang++ -MD -MT build/dev/transport/protocol_server.hh.o -MF build/dev/transport/protocol_server.hh.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=9 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=21 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -fextend-variable-liveness=none -march=westmere -I/usr/include/p11-kit-1 -DBOOST_ALL_DYN_LINK    -fvisibility-inlines-hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -fextend-variable-liveness=none -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -fpch-validate-input-files-content -O2 -DSCYLLA_BUILD_MODE=dev -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/dev/gen  --include transport/protocol_server.hh -c -o build/dev/transport/protocol_server.hh.o build/dev/gen/empty.cc -USCYLLA_USE_PRECOMPILED_HEADER
16:31:09  In file included from <built-in>:1:
16:31:09  ./transport/protocol_server.hh:46:42: error: use of undeclared identifier 'foreign_ptr'
16:31:09     46 |     virtual future<utils::chunked_vector<foreign_ptr<std::unique_ptr<client_data>>>> get_client_data() {
16:31:09        |                                          ^~~~~~~~~~~
16:31:09  ./transport/protocol_server.hh:47:56: error: use of undeclared identifier 'foreign_ptr'
16:31:09     47 |         return make_ready_future<utils::chunked_vector<foreign_ptr<std::unique_ptr<client_data>>>>();
16:31:09        |                                                        ^~~~~~~~~~~
16:31:09  2 errors generated.
```

Finally something that makes some sense... :D

vladzcloudius commented:
@avikivity @ruthea CI passed. We are good to go into a review.

avikivity commented:
> Thanks for your work on this @vladzcloudius .
> 
> @avikivity - looks like there is a simple fix vlad is working on to ensure CI/CD test success. Everything else look good from your perspective? Our goal here is to get it for sure in 2026.1 and perhaps even backported to 2025.4 (per Dor's suggestion)... but it all starts with a commit to master. Thanks for your help!

Will only backport with a notarized release absolving and indemnifying me from any regressions

avikivity commented on client_data.hh:
```diff
@@ -46,6 +53,7 @@ struct client_data {
     std::optional<sstring> ssl_protocol;
     std::optional<sstring> username;
     std::optional<sstring> scheduling_group_name;
+    std::optional<std::list<std::pair<client_options_cache_entry_type, client_options_cache_entry_type>>> client_options;
```
btw they keys might as well be an sstring, it doesn't allocate for strings <= 30 chars.

vladzcloudius replied:
> I now see why it's optional, it's to distinguish between a client that sent options and a client that did not.


`cql_server::connection::process_startup(...)` will raise an exception if OPTIONS is not provided AFAICT:

```
future<std::unique_ptr<cql_server::response>> cql_server::connection::process_startup(uint16_t stream, request_reader in, service::client_state& client_state,
        tracing::trace_state_ptr trace_state) {
    utils::result_with_exception_ptr<std::unordered_map<sstring, sstring>> o = in.read_string_map();
    if (!o) {
        co_return coroutine::exception(std::move(o).assume_error());
    }
    std::unordered_map<sstring, sstring> options = std::move(o).assume_value();
...
```

And this seem to follow the binary protocol (v5 at least) definition: https://cassandra.apache.org/doc/latest/cassandra/_attachments/native_protocol_v5.html (See "4.1.1 STARTUP").

So, again, IMO you are right and dropping the "optional" for `client_state._client_options` makes sense.

avikivity commented:
Looks generally good, had some comments.

vladzcloudius commented on test/cqlpy/test_virtual_tables.py:
```diff
@@ -76,6 +76,7 @@ def test_clients(scylla_only, cql):
         'ssl_enabled',
         'ssl_protocol',
         'username',
+        'client_options'
```
Will add. NP.

vladzcloudius replied:
done

vladzcloudius commented on alternator/server.cc:
```diff
@@ -708,8 +708,12 @@ future<executor::request_return_type> server::handle_api_request(std::unique_ptr
+
     auto system_clients_entry = _ongoing_requests.emplace(
-        req->get_client_address(), req->get_header("User-Agent"),
+        req->get_client_address(), std::move(user_agent_header),
```
First of all - there is no such guarantee. 
And I don't understand which optimization you keep on referencing.

vladzcloudius commented on transport/server.cc:
```diff
@@ -1619,6 +1619,9 @@ std::unique_ptr<cql_server::response> cql_server::connection::make_supported(int
     opts.insert({"COMPRESSION", "snappy"});
+    // CLIENT_OPTIONS value is a JSON string that can be used to pass client-specific configuration,
+    // e.g. CQL driver configuration.
+    opts.insert({"CLIENT_OPTIONS", ""});
```
done

vladzcloudius commented:
New in v3:
- Drop `std::optional` from `client_state._client_options` and `client_data`.
- Added coma after the last entry of the list used in a test.
- Move the patch that adds new `CLIENT_OPTIONS` option declaration to the end of the PR. 
- Define and use a dedicated empty string type instead of `empty_t`.
- Added a entry describing `CLIENT_OPTIONS` in the `docs/dev/protocol-extensions.md`
